### PR TITLE
5주차 과제 제출 - 신명희

### PR DESCRIPTION
**Pull request by mh0223**
### 코드 작성 체크포인트

- 가독성
    - [x] 변수와 함수의 이름이 용도와 기능을 명확히 나타내는가?
    - [x] 중복되는 로직을 별도의 함수로 추출했는가?
    - [x] 중첩된 if 문을 피하고 조건문을 단순화했는가?
    - [x] 매직 넘버와 반복되는 문자열을 상수로 정의했는가?
    - [x] 데이터 처리에 reduce, filter, map 등의 고차 함수를 사용했는가?
    - [x] 일관된 코딩 스타일, 적절한 들여쓰기, 논리적인 함수 배치로 가독성을 향상시켰는가?
- 단일 책임 원칙
    - [x]  함수/훅이 하나의 명확한 목적만을 가지고 있는가?
    - [x]  함수/훅의 이름이 그 목적을 명확하게 나타내는가?
- 순수 함수
    - [x]  동일한 입력에 대해 항상 동일한 출력을 반환하는가?
    - [x]  외부 상태를 변경하지 않는가?
    - [ ]  입력 데이터를 직접 수정하지 않고 새로운 데이터를 반환하는가?
- 합성 가능성
    - [x]  함수/훅을 다른 함수/훅과 쉽게 조합할 수 있는가?
- 관심사 분리
    - [x]  비즈니스 로직과 UI 로직이 적절히 분리되어 있는가?
- DRY (Don't Repeat Yourself) 원칙
    - [x]  이 함수/훅이 중복 코드를 제거하는 데 도움이 되는가?
- 테스트 용이성
    - [x]  함수/훅에 대한 단위 테스트를 쉽게 작성할 수 있는가?
    (input에 대한 output을 정의하기 수월한가)

### advanced 과제 체크포인트

- [x] advanced.test.ts에 있는 시나리오 테스트를 모두 통과한다.
- [x] 3개 이상의 custom hook을 통해서 리팩토링을 진행했고, 해당 hook에 대한 테스트 코드가 작성되었다.
- [x] 3개 이상의 순수함수를 통해서 리팩토링을 진행했고, 해당 함수에 대한 테스트 코드가 작성되었다.

### 고민했던 부분
- 
- 

### 새롭게 알게 된 점
- 
- 

## 해결하지 못한 문제

- 
- 

## 리뷰 요청 사항

- 
- 

### 성장 포인트

1.
2.
3.

### 개선하고 싶은 부분
- 
-

## 인사이트  

현재 작성된 코드가 요구사항에 잘 대응할 수 있는지 자체적으로 점검해보세요. **내가 작성한 코드가 좋은 코드인지 판단해볼 수 있습니다.**

1. start:advanced로 실행하면 초기 데이터를 localStorage에서 불러오고, 추가하고, 수정하고, 삭제할 수 있다.
2. start:api-mock로 실행하면 초기 데이터를 API를 통해서 불러오고, 추가하고, 수정하고, 삭제할 수 있다.
    - API는 [msw](https://mswjs.io/)로 만든다.
3. 회원 등급별 할인 기능
    - 관리자는 회원 등급을 추가할 수 있다.
    - 회원 등급에 따른 할인율이 존재한다.
        - 예시) 일반: 3%, 실버: 5%, 골드: 7%, VIP: 10%,
    - 장바구니 페이지에서 현재 회원 등급을 임의로 선택할 수 있다.
    - 회원 등급에 따라 총 구매 가격에 할인율이 적용되어야 한다.

이 외에도 다양한 요구사항이 추가될 수 있습니다.

### CODE REVIEW

**Comment by mh0223** [comment]
> Important
> 
> ## Review skipped
> More than 25% of the files skipped due to max files limit. The review is being skipped to prevent a low-quality review.
> 
> 24 files out of 80 files are above the max files limit of 50. Please upgrade to Pro plan to get higher limits.
> 
> You can disable this status message by setting the `reviews.review_status` to `false` in the CodeRabbit configuration file.
> 
> Thank you for using CodeRabbit. We offer it for free to the OSS community and would appreciate your support in helping us grow. If you find it useful, would you consider giving us a shout-out on your favorite social media?
> 
> Share
> Tips

피드백 다시 해줘

**Comment by mh0223** [comment]
@coderabbitai PR 리뷰 새로운 커밋들 반영해서 다시 해줘

**Comment by mh0223** [comment]
@coderabbitai review

**Comment by mh0223** [comment]
@coderabbitai review

**Comment by mh0223** [comment]
@coderabbitai  리뷰를 좀 자세히 해주면 안될까..

**Comment by mh0223** [comment]
@coderabbitai 내 코드에서 개선할 점을 말해줘

**Comment by mh0223** [comment]
> @mh0223: Here are some areas for improvement based on the detailed review of your code:
> 
> 1. **Code Duplication**:
>    
>    * There are several instances of code duplication, especially in the `CartPage` and `AdminPage` components across different directories (`origin`, `refactoring`, and `fsd`). Consider creating reusable components and hooks to avoid redundancy.
> 2. **Consistent Naming Conventions**:
>    
>    * Ensure consistent naming conventions across your components, hooks, and functions. For example, in some places, you use camelCase (`handleAddDiscount`), while in others, you use snake_case (`calculate_item_total`). Stick to one convention for better readability.
> 3. **Error Handling**:
>    
>    * Implement error handling in your API calls and state updates. For instance, when updating product quantities or applying coupons, ensure you handle potential errors gracefully to improve user experience.
> 4. **TypeScript Types**:
>    
>    * Enhance the use of TypeScript types to ensure type safety. For example, define more specific types for your state and props to avoid any type-related issues. This can also help with better code completion and error detection in your IDE.
> 5. **Optimizing State Management**:
>    
>    * In your state management using Zustand, consider using selectors to optimize performance. This can help avoid unnecessary re-renders by only selecting the specific state slices needed by your components.
> 6. **Unit Tests**:
>    
>    * While you have included unit tests, ensure that all edge cases are covered. For instance, test scenarios where the cart is empty, where products are out of stock, and where invalid coupon codes are applied.
> 7. **Code Comments and Documentation**:
>    
>    * Add comments and documentation to your code, especially for complex logic and custom hooks. This will help other developers (and your future self) understand the code better and maintain it more easily.
> 8. **CSS and Styling**:
>    
>    * Ensure consistent styling across your components. You are using Tailwind CSS, so make sure to follow a consistent pattern for class names and avoid inline styles where possible.
> 9. **Performance Optimization**:
>    
>    * Look for opportunities to optimize performance, such as memoizing expensive calculations and using React's `useMemo` and `useCallback` hooks to prevent unnecessary re-renders.
> 10. **Code Structure and Organization**:
>     
>     * Organize your code structure to follow a clear and logical pattern. Group related files together and use index files to simplify imports. This will make your codebase more navigable and easier to understand.
> 
> By addressing these areas, you can improve the overall quality, maintainability, and performance of your code. If you need further assistance or specific examples, feel free to ask!
> 
> Analysis chain
> Scripts executed
> Actions performed

한국어로 말해줄래

**Comment by mh0223** [comment]
> @mh0223: Here are some areas for improvement based on the detailed review of your code:
> 
> 1. **Code Duplication**:
>    
>    * There are several instances of code duplication, especially in the `CartPage` and `AdminPage` components across different directories (`origin`, `refactoring`, and `fsd`). Consider creating reusable components and hooks to avoid redundancy.
> 2. **Consistent Naming Conventions**:
>    
>    * Ensure consistent naming conventions across your components, hooks, and functions. For example, in some places, you use camelCase (`handleAddDiscount`), while in others, you use snake_case (`calculate_item_total`). Stick to one convention for better readability.
> 3. **Error Handling**:
>    
>    * Implement error handling in your API calls and state updates. For instance, when updating product quantities or applying coupons, ensure you handle potential errors gracefully to improve user experience.
> 4. **TypeScript Types**:
>    
>    * Enhance the use of TypeScript types to ensure type safety. For example, define more specific types for your state and props to avoid any type-related issues. This can also help with better code completion and error detection in your IDE.
> 5. **Optimizing State Management**:
>    
>    * In your state management using Zustand, consider using selectors to optimize performance. This can help avoid unnecessary re-renders by only selecting the specific state slices needed by your components.
> 6. **Unit Tests**:
>    
>    * While you have included unit tests, ensure that all edge cases are covered. For instance, test scenarios where the cart is empty, where products are out of stock, and where invalid coupon codes are applied.
> 7. **Code Comments and Documentation**:
>    
>    * Add comments and documentation to your code, especially for complex logic and custom hooks. This will help other developers (and your future self) understand the code better and maintain it more easily.
> 8. **CSS and Styling**:
>    
>    * Ensure consistent styling across your components. You are using Tailwind CSS, so make sure to follow a consistent pattern for class names and avoid inline styles where possible.
> 9. **Performance Optimization**:
>    
>    * Look for opportunities to optimize performance, such as memoizing expensive calculations and using React's `useMemo` and `useCallback` hooks to prevent unnecessary re-renders.
> 10. **Code Structure and Organization**:
>     
>     * Organize your code structure to follow a clear and logical pattern. Group related files together and use index files to simplify imports. This will make your codebase more navigable and easier to understand.
> 
> By addressing these areas, you can improve the overall quality, maintainability, and performance of your code. If you need further assistance or specific examples, feel free to ask!
> 
> Analysis chain
> Scripts executed
> Actions performed

@coderabbitai 한국어로 말해줄래

**Review by jane1107 — COMMENTED** [review]
안녕하세요, 프론트 1기 수료생 한현정입니다. 기술적 인사이트를 주고 받기 위해 피어리뷰 남겨드립니다. 변수의 이름이 명확하게 표현되었고, fsd 구조에 맞게 모듈화가 잘 된 것 같습니다. 
상수를 적극적으로 사용하고, 코딩스타일이 일관되어 가독성이 좋았습니다. 필요한 경우 고차함수를 잘 사용해주신 것 같습니다!!

**Comment by mh0223** [comment]
> 안녕하세요, 프론트 1기 수료생 한현정입니다. 기술적 인사이트를 주고 받기 위해 피어리뷰 남겨드립니다. 변수의 이름이 명확하게 표현되었고, fsd 구조에 맞게 모듈화가 잘 된 것 같습니다. 상수를 적극적으로 사용하고, 코딩스타일이 일관되어 가독성이 좋았습니다. 필요한 경우 고차함수를 잘 사용해주신 것 같습니다!!

피드백 감사합니다!! : D

File: packages/assignment-5/src/fsd/entities/cart/cart.model.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+          );
+        }
+
+        const cartItem = state.cartItems.find(
+          (cartItem) => cartItem.product.id === targetProductId
+        );
+
+        if (cartItem) {
+          let validNewQuantity = Math.max(MIN_QUANTITY, newQuantity);
+          validNewQuantity = Math.min(cartItem.product.stock, validNewQuantity);
```
Comment (jung-han) [inline packages/assignment-5/src/fsd/entities/cart/cart.model.ts:63]: 이름을 다르게 지어서 const로 사용해보세용

File: packages/assignment-5/src/fsd/entities/coupon/coupon.const.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import { Coupon } from './coupon.types';
+
+export const INIT_NEW_COUPON: Coupon = {
```
Comment (jung-han) [inline packages/assignment-5/src/fsd/entities/coupon/coupon.const.ts:3]: 보통 init이란 표현은 메서드이름에 많이 사용하니 defaultCouponValue 나 defaultCouponFormValue
요런식으로 작성하면 좋을것 가타요

File: packages/assignment-5/src/fsd/entities/coupon/coupon.model.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
+import { produce } from 'immer';
+import { create } from 'zustand';
+import { type Coupon } from '~entities/coupon';
+
+export const INIT_COUPONS: Coupon[] = [
```
Comment (jung-han) [inline packages/assignment-5/src/fsd/entities/coupon/coupon.model.ts:5]: 요기두용

File: packages/assignment-5/src/fsd/entities/coupon/coupon.model.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    name: '10% 할인 쿠폰',
+    code: 'PERCENT10',
+    discountType: 'percentage',
+    discountValue: 10,
+  },
+];
+
+type State = {
+  coupons: Coupon[];
+  selectedCoupon: Coupon | null;
```
Comment (jung-han) [inline packages/assignment-5/src/fsd/entities/coupon/coupon.model.ts:22]: `| null` 을 하는 함수들이 많아지면 그냥 `Nullable<T>` 타입을 하나 유틸성으로 선언해서 사용하셔도 가독성도 좋아지고 훨씬 편하실 것 같아요

File: packages/assignment-5/src/types.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+export interface CartItem {
+  product: Product;
+  quantity: number;
+}
+
+export interface Coupon {
+  name: string;
+  code: string;
+  discountType: 'amount' | 'percentage';
```
Comment (jung-han) [inline packages/assignment-5/src/types.ts:22]: 'amount' | 'percentage'
요거도 지금 검색해봤을때는 여러군데에서 쓰고 있어서 타입을 분리해서 그걸 import해서 쓰는게 좋을것 같아여

File: packages/assignment-5/.eslintrc.cjs [hunk 6]
```diff
@@ -0,0 +1,29 @@
+module.exports = {
+  root: true,
+  env: { browser: true, es2020: true },
+  extends: [
+    "eslint:recommended",
+    "plugin:@typescript-eslint/recommended",
+    "plugin:react-hooks/recommended",
+    "plugin:prettier/recommended",
```
Comment (jung-han) [inline packages/assignment-5/.eslintrc.cjs:8]: 멘토링때말했던것처럼 fsd쓰면 꼭 린트룰 같이 쓰세요~

File: packages/assignment-5/src/fsd/features/cart/cart-update-form/cart-update-form.model.ts [hunk 7]
```diff
@@ -0,0 +1,14 @@
+import { CartItem } from '~entities/cart';
+
+export const calculateAppliedDiscount = (cartItem: CartItem) => {
+  const { discounts } = cartItem.product;
+  const { quantity } = cartItem;
+
+  let appliedDiscount = 0;
+  for (const discount of discounts) {
```
Comment (jung-han) [inline packages/assignment-5/src/fsd/features/cart/cart-update-form/cart-update-form.model.ts:8]: reduce를 활용할 수 있겠어여

File: packages/assignment-5/src/fsd/features/product/product-create-form/product-create-form.model.tsx [hunk 8]
```diff
@@ -0,0 +1,11 @@
+export const handleAddNewProduct = () => {
+  const productWithId = { ...newProduct, id: Date.now().toString() };
+  onProductAdd(productWithId);
```
Comment (jung-han) [inline packages/assignment-5/src/fsd/features/product/product-create-form/product-create-form.model.tsx:3]: 요 파일 에러 안나요?

File: packages/assignment-5/src/fsd/features/product/product-create-form/product-create-form.ui.tsx [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+    addProduct(newProduct);
+    onToggleNewForm();
+  };
+
+  const handleInputChange = (
+    e: ChangeEvent<HTMLInputElement>,
+    type: keyof Product
+  ) => {
+    const newValue =
+      e.target.type === 'number' ? parseInt(e.target.value) : e.target.value;
```
Comment (jung-han) [inline packages/assignment-5/src/fsd/features/product/product-create-form/product-create-form.ui.tsx:28]: 요런거 반복적이게 있는 것 같은데 유틸성 메서드로 분리해서 써도 되겠네여

File: packages/assignment-5/src/fsd/features/product/product-list/product-list-model.tsx [hunk 10]
```diff
@@ -0,0 +1,13 @@
+import { useStore } from 'zustand';
+import { cartStore } from '~entities/cart';
+import { productStore, type Product } from '~entities/product';
+
+export const getMaxDiscount = ({ discounts }: Product) => {
+  return discounts.reduce((max, discount) => Math.max(max, discount.rate), 0);
+};
+
+export const calculateRemainingStock = (product: Product) => {
+  const cartItems = useStore(cartStore, (s) => s.cartItems);
```
Comment (jung-han) [inline packages/assignment-5/src/fsd/features/product/product-list/product-list-model.tsx:10]: 종종 s, p 이렇게 작성하는 경우들이 있는데, 개인적으로는 풀어서 명확하게 그게 뭔지 작성하는게 좋아보여요
앞에서도 복수형을 사용하고 있기때문에 이름을 뒤에서 단수형으로 쓰는게 그렇게 어려운 건 아니니까용

File: packages/assignment-5/src/fsd/pages/cart/cart-page.model.ts [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    } else {
+      couponDiscount =
+        ((cartTotalPrice - quantityDiscount) * selectedCoupon.discountValue) /
+        100;
+    }
+  }
+
+  const calculateTotalPrice =
+    cartTotalPrice - quantityDiscount - couponDiscount;
+
```
Comment (jung-han) [inline packages/assignment-5/src/fsd/pages/cart/cart-page.model.ts:59]: 멘토링때 말씀드린것처럼 이렇게 절차적으로 쭉 작성할수도 있지만, discount를 계산해주는 비즈니스 로직이 모인 순수함수를 하나 작성해서 사용한다면 가독성이 좋아질수도 있을 것 같아요!

File: packages/assignment-5/src/refactoring/hooks/utils/cartUtils.ts [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+  return appliedDiscount;
+};
+
+export const calculateCartTotal = (
+  cart: CartItem[],
+  selectedCoupon: Coupon | null
+) => {
+  let beforeDiscountPrice = 0;
+  beforeDiscountPrice = cart.reduce((sumCartPrice, currentItem) => {
+    sumCartPrice += currentItem.product.price * currentItem.quantity; // 창기님이 reduce에서 return으로 바로 계산식을 반환하기 보다는 이렇게 빼서 다 더해진 sum 값만 리턴하면 깔끔하다고 알려주셨음
```
Comment (jung-han) [inline packages/assignment-5/src/refactoring/hooks/utils/cartUtils.ts:33]: _(복잡한 계산일때만)_

File: packages/assignment-5/src/refactoring/hooks/utils/cartUtils.ts [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+
+  let afterDiscountPrice = beforeDiscountPrice - itemDiscountPrice;
+
+  let couponDiscountPrice = 0;
+  if (selectedCoupon) {
+    couponDiscountPrice =
+      selectedCoupon.discountType === "amount"
+        ? selectedCoupon.discountValue
+        : (afterDiscountPrice * selectedCoupon.discountValue) / 100;
+    afterDiscountPrice -= couponDiscountPrice;
```
Comment (jung-han) [inline packages/assignment-5/src/refactoring/hooks/utils/cartUtils.ts:53]: let으로 두지 않아도 될 것 같아여

**Review by jung-han — COMMENTED** [review]
간단한 피드백 남겼습니다~ 
FSD라니 다른 분들보다 훨씬 노력을 많이 해주셨네요 👍 
앞으로도 잘 부탁드립니다

File: packages/assignment-5/src/refactoring/hooks/utils/cartUtils.ts (hunk 13 above)
Comment (mh0223) [reply packages/assignment-5/src/refactoring/hooks/utils/cartUtils.ts:53]: 피드백 감사합니다!! : D
